### DISCUSSION TITLE
Render hdri in ImGUI Image (in viewport)

### DISCUSSION THREAD

**Discussion by jayanam** [ocornut/imgui#3831] [Old/Archived discussions]
Hi guys,
Im using the docking branch and ImGUI with opengl backend. I am rendering opengl into a framebuffer and then add this to ImGui::Image, work pretty nice:

![image](https://user-images.githubusercontent.com/18039584/108737155-9a271800-7532-11eb-99b9-4853a666b9f4.png)

```
uint64_t textureID = mFrameBuffer->get_texture();
// I know reinterpret_cast is evil - better solution?
ImGui::Image(reinterpret_cast<void*>(textureID), ImVec2{ mSize.x, mSize.y }, ImVec2{ 0, 1 }, ImVec2{ 1, 0 });
```
I use a simple directional light fragment shader.

My question is: If you want to light the scene with an hdri (different shader), how can I combine this with the framebuffer? Does someone has an example or can tell me the basic steps, just to be on the right track.

Thx in advance!
Jay

**Comment by PathogenDavid** [comment]
> My question is: If you want to light the scene with an hdri (different shader), how can I combine this with the framebuffer? Does someone has an example or can tell me the basic steps, just to be on the right track.

This strikes me as a question outside of the scope of Dear ImGui. ImGui doesn't care about how you render your scene.

> ```cpp
> // I know reinterpret_cast is evil - better solution?
> ```

This is a sensible use of `reinterpret_cast`, but it is worth noting that this isn't 32-bit portable if that's something you care about.

**Comment by ocornut** [comment]
I think the crux of the question is:
> If you want to light the scene with an hdri (different shader),

Some answers:

- You can use `ImDrawList::AddCallback()` to communicate state to your renderer, e.g. add a callback with a data requesting shader change.  A game I wrote (https://www.thedragonstrap.com) was entirely rendering via ImDrawList + callback to manipulate shader, blending state.
- Note that there's a helper you might or not use: passing `ImDrawCallback_ResetRenderState` as a callback is a standardized way of requesting the backend to restore all restore state (shader state). You may do that or restore manually only what you modified using a callback.
- In your own engine (and custom renderer backend), another way is to use `ImTextureID` to express a higher level concept e.g. a material which bundles a texture and the way to render it.

> I know reinterpret_cast is evil - better solution?

Solution is to set your mind free from dogmas :)

> worth noting that this isn't 32-bit portable 

Hmm.. Nothing strikes me as not portable there?

**Reply by jayanam** [reply]
Thx for the answers, helps a lot!
> Solution is to set your mind free from dogmas :)

Dont worry, my mind is free from them... but I ran into issues for applications compiling for 32/64 bit :-)

**Reply by PathogenDavid** [reply]
> Hmm.. Nothing strikes me as not portable there?

Their texture ID is a `uint64_t` which has to be truncated to fit into a 32-bit `void*`

**Reply by ocornut** [reply]
Oh I see. You can `#define ImTextureID uint64_t` in your config file to fix it.
